🕊️ DPO padding free

qgallouedec opened the pull request:
# What does this PR do?

demo; further experiments in next comment

```python
from datasets import load_dataset
from transformers import AutoModelForCausalLM, AutoTokenizer
from trl import DPOConfig, DPOTrainer
import torch

model_id = "Qwen/Qwen2-0.5B-Instruct"
model = AutoModelForCausalLM.from_pretrained(model_id, torch_dtype=torch.bfloat16, attn_implementation="flash_attention_2")
tokenizer = AutoTokenizer.from_pretrained(model_id)
dataset = load_dataset("trl-lib/ultrafeedback_binarized", split="train[:10%]")
training_args = DPOConfig(output_dir="Gemma2-2B-DPO-pf", max_prompt_length=128, max_completion_length=128, logging_steps=10, padding_free=True)
trainer = DPOTrainer(model=model, args=training_args, train_dataset=dataset, processing_class=tokenizer)
trainer.train()
```

With and without padding-free (not sure why they don't match exactly, the logits do precisely match though)
<img width="1042" alt="Screenshot 2024-12-26 at 22 00 06" src="https://github.com/user-attachments/assets/b55cc47c-a8c4-4033-9bc5-d54d64f21a3e" />


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/trl/tree/main/docs).
- [ ] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/trl/pr_2520). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

qgallouedec commented on .github/workflows/tests.yml:
```diff
@@ -89,6 +89,7 @@ jobs:
           python -m pip install -U git+https://github.com/huggingface/datasets.git
           python -m pip install -U git+https://github.com/huggingface/transformers.git
           python -m pip install ".[dev]"
+
```
```suggestion
```

qgallouedec commented on trl/trainer/dpo_trainer.py:
```diff
@@ -1149,12 +1158,22 @@ def concatenated_forward(self, model: nn.Module, batch: dict[str, Union[list, to
 
-            outputs = model(input_ids=input_ids, attention_mask=attention_mask, **model_kwargs)
+            if self.padding_free:
+                # Flatten the input_ids, position_ids, and loss_mask
+                # input_ids = [[a, b, c, 0], ->     input_ids = [a, b, c, d, e, f, g]
+                #              [d, e, f, g]]     position_ids = [0, 1, 2, 0, 1, 2, 3]
```
```suggestion
                # Flatten the input_ids, position_ids, and loss_mask
                # input_ids = [[a, b, c, 0], ->     input_ids = [[a, b, c, d, e, f, g]]
                #              [d, e, f, g]]     position_ids = [[0, 1, 2, 0, 1, 2, 3]]
```

qgallouedec commented on trl/trainer/dpo_config.py:
```diff
@@ -145,6 +145,9 @@ class DPOConfig(TrainingArguments):
             useful for saving memory and speeding up training by not computing the logits for all tokens, especially in
             scenarios when working with very long prompts where labels are ignored (-100).
             [Read more](https://huggingface.co/docs/transformers/main/model_doc/llama#transformers.LlamaForCausalLM)
+        padding_free (`bool`, *optional*, defaults to `False`):
+            Whether the forward passes are performed without padding, i.e. flattening all the samples in the batch
+            into a single sample, associated with a position_ids vector. Only possible with flash-attention.
```
```suggestion
        padding_free (`bool`, *optional*, defaults to `False`):
            Whether forward passes are performed without padding by flattening all sequences in the batch 
            into a single continuous sequence. This approach requires associating a `position_ids` vector to track 
            positional information. Currently, this is only supported with the `flash_attention_2` mechanism, as it 
            can handle the flattened batch structure.
```

qgallouedec commented on trl/trainer/dpo_trainer.py:
```diff
@@ -1152,12 +1164,22 @@ def concatenated_forward(self, model: nn.Module, batch: dict[str, Union[list, to
+                #              [d, e, f, g]]     position_ids = [[0, 1, 2, 0, 1, 2, 3]]
+                input_ids = input_ids[attention_mask.bool()].unsqueeze(0)
+                loss_mask = loss_mask[attention_mask.bool()].unsqueeze(0)
+                model_kwargs["position_ids"] = attention_mask.cumsum(1)[attention_mask.bool()].unsqueeze(0) - 1
```
```suggestion
                position_ids = attention_mask.cumsum(1)[attention_mask.bool()].unsqueeze(0) - 1
                model_kwargs["position_ids"] = position_ids
```

qgallouedec commented on trl/trainer/dpo_trainer.py:
```diff
@@ -1208,8 +1241,22 @@ def concatenated_forward(self, model: nn.Module, batch: dict[str, Union[list, to
+            # There are 2*num_examples ranges in total: the first half corresponds to the chosen tokens,
+            # and the second half to the rejected tokens.
+            # To find the start of the rejected tokens, we look for the num_examples+1-th zero in pos_id.
+            split_idx = (model_kwargs["position_ids"] == 0).nonzero(as_tuple=True)[1][num_examples]
```
```suggestion
            split_idx = (position_ids == 0).nonzero(as_tuple=True)[1][num_examples]
```

qgallouedec commented:
## Regression test:

```python
from datasets import load_dataset
from transformers import AutoModelForCausalLM, AutoTokenizer
from trl import DPOConfig, DPOTrainer
import torch

model_id = "Qwen/Qwen2-0.5B-Instruct"
model = AutoModelForCausalLM.from_pretrained(model_id, torch_dtype=torch.bfloat16, attn_implementation="flash_attention_2")
tokenizer = AutoTokenizer.from_pretrained(model_id)
dataset = load_dataset("trl-lib/ultrafeedback_binarized", split="train[:10%]")
# dataset = load_dataset("trl-internal-testing/zen", "standard_preference", split="train")
training_args = DPOConfig(output_dir="Qwen2-0.5B-DPO-no_pf", max_prompt_length=128, max_completion_length=128, logging_steps=10, padding_free=False)
trainer = DPOTrainer(model=model, args=training_args, train_dataset=dataset, processing_class=tokenizer)
trainer.train()
```

Is the new `padding_free=False` (`no_pf` in screenshot) equivalent to DPO on current main branch (`main` in screenshot)? -> yes

<img width="2132" alt="Screenshot 2025-01-07 at 20 15 31" src="https://github.com/user-attachments/assets/f2019381-722d-46bb-8258-6a99b75861d8" />

Does `padding_free=True` (`pf` in screenshot) results match `padding_free=False` (`no_pf` in screenshot) results? -> Yes

<img width="2132" alt="Screenshot 2025-01-07 at 20 19 41" src="https://github.com/user-attachments/assets/4920afd1-92e8-415e-80fd-655604ef45bf" />

(note: on screenshots its written "Gemma" but it's actually a Qwen model trained)

qgallouedec commented on trl/trainer/dpo_trainer.py:
```diff
@@ -1149,15 +1161,26 @@ def concatenated_forward(self, model: nn.Module, batch: dict[str, Union[list, to
+            logits = outputs.logits
 
             # Offset the logits by one to align with the labels
-            logits = outputs.logits[:, :-1, :]
-            labels = input_ids[:, 1:].clone()
-            loss_mask = loss_mask[:, 1:].bool()
```
rolling works in both cases (flattened tensors and batched)

```python
# input_ids = [[1, 2, 3, 4],
#              [5, 6, 7, 8]]
# loss_mask = [[0, 0, 1, 1],
#              [0, 0, 0, 1]]
labels = input_ids[:, 1:].clone()
loss_mask = loss_mask[:, 1:].bool()
# labels =    [[2, 3, 4],
#              [6, 7, 8]]
# loss_mask = [[0, 1, 1],
#              [0, 0, 1]]

# But

# input_ids = [[1, 2, 3, 4, 5, 6, 7, 8]]
# loss_mask = [[0, 0, 1, 1, 0, 0, 0, 1]]
labels = input_ids[:, 1:].clone()
loss_mask = loss_mask[:, 1:].bool()
# labels =    [[2, 3, 4, 5, 6, 7, 8]]
# loss_mask = [[0, 1, 1, 0, 0, 0, 1]]
```

The first token of the first sequence (1) is removed but not the first token of the second sequence (5). To align the labels while keeping a consistent behaviour across sequences in the batch, we use roll instead. The only difference is that the first token, instead of being discarded, is appended to the end.

August-murr reviewed (approved):
[LGTM!](https://youtu.be/rR4n-0KYeKQ?t=30)
